firedrake/slate/slac.py

```diff
@@ -0,0 +1,473 @@
+
+from singledispatch import singledispatch
+
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.form import Form
+
+
+__all__ = ['compile_slate_expression']
+
+
+class Transformer(Visitor):
+    """Replaces all out-put tensor references with a specified
+    name of :type: `Eigen::Matrix` with appropriate shape.
+
+    The default name of :data:`"A"` is assigned, otherwise a
+    specified name may be passed as the :data:`name` keyword
+    argument when calling the visitor."""
+
+    def visit_object(self, o, *args, **kwargs):
+        """Visits a a COFFEE object and returns it.
+
+        i.e. string ---> string
+        """
+        return o
+
+    def visit_list(self, o, *args, **kwargs):
+        """Visits an input of COFFEE objects and returns
+        the complete list of said objects."""
+
+        newlist = [self.visit(e, *args, **kwargs) for e in o]
+        if all(newo is e for newo, e in zip(newlist, o)):
+            return o
+
+        return newlist
+
+    visit_Node = Visitor.maybe_reconstruct
+
+    def visit_FunDecl(self, o, *args, **kwargs):
+        """Visits a COFFEE FunDecl object and reconstructs
+        the FunDecl body and header to generate
+        Eigen::MatrixBase C++ template functions.
+
+        Creates a template function for each subkernel form
+        template <typename Derived>:
+
+        template <typename Derived>
+        static inline void foo(Eigen::MatrixBase<Derived> const & A, ...)
+        {
+          [Body...]
+        }
+        """
+
+        name = kwargs.get("name", "A")
+        new = self.visit_Node(o, *args, **kwargs)
+        ops, okwargs = new.operands()
+        if all(new is old for new, old in zip(ops, o.operands()[0])):
+            return o
+
+        template = "template <typename Derived>"
+        pred = ["static", "inline"]
+        body = ops[3]
+        args, _ = body.operands()
+        nargs = [ast.FlatBlock("Eigen::MatrixBase<Derived> & %s = const_cast<Eigen::MatrixBase<Derived> &>(%s_);\n" % (name, name))] + args
+        ops[3] = nargs
+        ops[4] = pred
+        ops[6] = template
+
+        return o.reconstruct(*ops, **okwargs)
+
+    def visit_Decl(self, o, *args, **kwargs):
+        """Visits a declared tensor and changes its type to
+        :template: result `Eigen::MatrixBase<Derived>`.
+
+        i.e. double A[n][m] ---> Eigen::MatrixBase<Derived> const &A_
+        """
+
+        name = kwargs.get("name", "A")
+        if o.sym.symbol != name:
+            return o
+        newtype = "Eigen::MatrixBase<Derived> const &"
+
+        return o.reconstruct(newtype, ast.Symbol("%s_" % name))
+
+    def visit_Symbol(self, o, *args, **kwargs):
+        """Visits a COFFEE symbol and redefines it as a FunCall object.
+
+        i.e. A[j][k] ---> A(j, k)
+        """
+
+        name = kwargs.get("name", "A")
+        if o.symbol != name:
+            return o
+        shape = o.rank
+
+        return ast.FunCall(ast.Symbol(name), *shape)
```

Comment:
I'm kind of worried that this might undermine COFFEE analyses for optimisations, but let me summon @FabioLuporini here for an expert opinion.